M.Talha Ramzan
---
2330-0141
---
BS AI
---
lab 03
---
ANN


In [20]:
import numpy as np
import tensorflow as tf
from tensorflow import keras

In [21]:
X = np.array([[0],[1],[2],[3],[4],[5]], dtype=float)
y = np.array([[0],[0],[0],[1],[1],[1]], dtype=float)

In [22]:
model = keras.Sequential([
    keras.layers.Dense(6, activation='sigmoid', input_shape=(1,)),
    keras.layers.Dense(1, activation='sigmoid')
])

In [23]:
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

In [24]:
model.fit(x, y, epochs=100, verbose=1)

Epoch 1/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 982ms/step - accuracy: 0.5000 - loss: 0.7056
Epoch 2/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 65ms/step - accuracy: 0.5000 - loss: 0.7050
Epoch 3/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 97ms/step - accuracy: 0.5000 - loss: 0.7044
Epoch 4/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 98ms/step - accuracy: 0.5000 - loss: 0.7038
Epoch 5/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 84ms/step - accuracy: 0.5000 - loss: 0.7033
Epoch 6/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 86ms/step - accuracy: 0.5000 - loss: 0.7027
Epoch 7/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 135ms/step - accuracy: 0.5000 - loss: 0.7021
Epoch 8/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 98ms/step - accuracy: 0.5000 - loss: 0.7015
Epoch 9/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 131ms/step - accuracy: 0.5000 - loss: 0.7009
Epoch 10/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 89ms/step - accuracy: 0.5000 - loss: 0.7003
Epoch 11/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 79ms/step - accuracy: 0.5000 - loss: 0.6997
Epoch 12/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 79ms/step - accuracy: 0.5000 -

In [25]:
prediction = model.predict(np.array([3], dtype=float))
print("Prediction for 3:", prediction)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 69ms/step
Prediction for 3: [[0.53110594]]


Task 2:
Implement the model using Tanh and compare its performance with ReLU.

In [26]:
model = keras.Sequential([
    keras.layers.Dense(6, activation='tanh', input_shape=(1,)),
    keras.layers.Dense(1, activation='sigmoid')
])

In [27]:
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

In [28]:
model.fit(X, y, epochs=100, verbose=1)

Epoch 1/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 921ms/step - accuracy: 0.6667 - loss: 0.6386
Epoch 2/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 59ms/step - accuracy: 0.6667 - loss: 0.6375
Epoch 3/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 58ms/step - accuracy: 0.6667 - loss: 0.6365
Epoch 4/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 55ms/step - accuracy: 0.6667 - loss: 0.6354
Epoch 5/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 56ms/step - accuracy: 0.6667 - loss: 0.6344
Epoch 6/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 64ms/step - accuracy: 0.6667 - loss: 0.6334
Epoch 7/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 55ms/step - accuracy: 0.6667 - loss: 0.6323
Epoch 8/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 56ms/step - accuracy: 0.6667 - loss: 0.6313
Epoch 9/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 67ms/step - accuracy: 0.6667 - loss: 0.6302
Epoch 10/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 64ms/step - accuracy: 0.6667 - loss: 0.6292
Epoch 11/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 54ms/step - accuracy: 0.6667 - loss: 0.6282
Epoch 12/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 59ms/step - accuracy: 0.6667 - l

In [30]:
prediction = model.predict(np.array([3], dtype=float))
print("Prediction for 3:", prediction)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 73ms/step
Prediction for 3: [[0.6572778]]


Task 3
Increase the number of neurons in the hidden layer from 6 to 10 and train the model using
---

In [31]:
model = keras.Sequential([
    keras.layers.Dense(10, activation='relu', input_shape=(1,)),
    keras.layers.Dense(1, activation='sigmoid')
])

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [32]:
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

In [33]:
model.fit(X, y, epochs=100, verbose=1)

Epoch 1/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 1s/step - accuracy: 0.6667 - loss: 0.6256
Epoch 2/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 58ms/step - accuracy: 0.6667 - loss: 0.6235
Epoch 3/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 58ms/step - accuracy: 0.6667 - loss: 0.6215
Epoch 4/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 56ms/step - accuracy: 0.6667 - loss: 0.6195
Epoch 5/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 62ms/step - accuracy: 0.6667 - loss: 0.6175
Epoch 6/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 64ms/step - accuracy: 0.6667 - loss: 0.6156
Epoch 7/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 57ms/step - accuracy: 0.6667 - loss: 0.6137
Epoch 8/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 58ms/step - accuracy: 0.6667 - loss: 0.6119
Epoch 9/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 59ms/step - accuracy: 0.6667 - loss: 0.6101
Epoch 10/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 61ms/step - accuracy: 0.6667 - loss: 0.6084
Epoch 11/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 60ms/step - accuracy: 0.6667 - loss: 0.6067
Epoch 12/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 57ms/step - accuracy: 0.6667 - loss

In [35]:
prediction = model.predict(np.array([3], dtype=float))
print("Prediction for 3:", prediction)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 142ms/step
Prediction for 3: [[0.7043593]]


Model 1 (ReLU – 6 Neurons)

Activation: relu

Final Loss: 0.30

Final Accuracy: 1.00 (100%)

✔ Loss decreased properly

✔ Accuracy increased to perfect

Prediction for input 3: 0.85

 Best performance

🔹 Model 2 (Sigmoid – 6 Neurons)

Activation: sigmoid

Final Loss: 0.6502

Final Accuracy: 0.8333 (83%)

 Good but slower learning than ReLU

🔹 Model 3 (Tanh – 6 Neurons)

Activation: tanh

Final Loss: 0.5429

Final Accuracy: 0.6667 (66%)

 Moderate performance

🔹 Model 4 (ReLU – 10 Neurons)

Activation: relu

Final Loss: 0.5177

Final Accuracy: 0.6667 (66%)

 More neurons but not better accuracy

 Training Speed (Fast vs Slow)
 Fast Training:

ReLU activation

Because ReLU does not suffer from vanishing gradient problem.

So Model 1 (ReLU – 6 neurons) trained fastest and best.

 Slow Training:

Sigmoid (slowest)

Because sigmoid compresses values between 0 and 1 → gradients become very small.

So Model 2 (Sigmoid) is slowest.

Medium Speed:

Tanh (faster than sigmoid, slower than relu)

Final Conclusion

Fastest + Best Accuracy → ReLU (6 neurons)

Slowest → Sigmoid

Increasing neurons does NOT always improve accuracy.

Activation function matters more than number of neurons.